crates/typst-ide/src/definition.rs

```diff
@@ -0,0 +1,243 @@
+use ecow::{eco_vec, EcoString};
+use typst::foundations::{Label, Module, Selector, Value};
+use typst::model::Document;
+use typst::syntax::ast::AstNode;
+use typst::syntax::{ast, FileId, LinkedNode, Side, Source, Span, SyntaxKind};
+use typst::World;
+
+use crate::analyze::analyze_import;
+use crate::{deref_target, named_items, DerefTarget, NamedItem};
+
+/// Find the definition of the item under the cursor.
+///
+/// Passing a `document` (from a previous compilation) is optional, but enhances
+/// the autocompletions. Label completions, for instance, are only generated
+/// when the document is available.
+pub fn definition(
+    world: &dyn World,
+    document: Option<&Document>,
+    source: &Source,
+    cursor: usize,
+    side: Side,
+) -> Option<Definition> {
+    let root = LinkedNode::new(source.root());
+    let leaf = root.leaf_at(cursor, side)?;
+
+    let target = deref_target(leaf.clone())?;
+
+    let mut use_site = match target {
+        DerefTarget::VarAccess(node) | DerefTarget::Callee(node) => node,
+        DerefTarget::IncludePath(path) | DerefTarget::ImportPath(path) => {
+            let import_item =
+                analyze_import(world, &path).and_then(|v| v.cast::<Module>().ok())?;
+            return Some(Definition::module(&import_item));
+        }
+        DerefTarget::LabelRef(r) => {
+            let ref_node = r.cast::<ast::Ref>()?.target();
+            let sel = Selector::Label(Label::new(ref_node));
+            let elem = document?.introspector.query_first(&sel)?;
+            let span = elem.span();
+
+            return Some(Definition {
+                kind: DefinitionKind::Label,
+                name: r.text().clone(),
+                value: elem.label().map(Value::Label),
+                span,
+                name_span: Span::detached(),
+            });
+        }
+        DerefTarget::Label(..) | DerefTarget::Code(..) => {
+            return None;
+        }
+    };
+
+    let mut paths = eco_vec![];
```

Comment:
Why do we collect the path here and continue just to early return below when the path is non-empty?

Reply:
I can continue my work on field definition and I don't think they cause any notable overhead, but I have removed them.